Fixes #35122 - Support gitlab oidc provider

j-m-harris opened the pull request.

ehelms commented:
[test foreman]

stejskalleos commented:
ok to test

stejskalleos commented:
Hi @j-m-harris, thanks for contributing to the Foreman,
I took quick look and have some points / questions:
* The integration test fails are not related to your changes, but failing `unit/sso/apache_test.rb` probably is
* `test/unit/sso/openid_connect_test.rb` should be also updated and cover these new changes
* Are there any changes in behavior that needs to be reflected in [docs - Configuring external auth admin](https://docs.theforeman.org/nightly/Administering_Red_Hat_Satellite/index-foreman-el.html#Configuring_External_Authentication_admin)? 
* Could you post instructions how to setup simple dev env and what are the testing steps for this pull request? 

If anything feel free to ping me on IRC (@lstejska)

j-m-harris commented:
Thanks for the feedback @stejskalleos, agreed the test coverage and documentation need updating for the gitlab oidc integration.

j-m-harris committed:
Fixes #35122 - Support gitlab oidc provider

j-m-harris commented:
@stejskalleos Fixed the test regression and added coverage for the new logic in openid_connect. Fixed some reversed assert_equal calls elsewhere in that test suite too.

Setting up gitlab as the oidc provider for foreman is similar to the keycloak integration. I can format the following into the foreman docs.

### Configure GitLab

* Create an Instance OAuth application.
   * Name: Foreman
   * Redirect URI: `https://foreman.example.com/users/extlogin/redirect_uri`
   * Trusted: False
   * Confidental: True
   * Scopes: openid, profile, email

### Configure foreman

* Go to Administer - Settings - Authentication
* Check 'Authorise login delegation auth source user autocreate' is set to 'External' (enables auto-creation of users from external OpenID provider).
* Set 'Authorise login delegation' to 'Yes'
* Set `OIDC Algorithm` to `RS256`
   * (id_token_signing_alg_values_supported from gitlab openid-configuration)
* Set `OIDC JWKs URL` to `https://gitlab.example.com/oauth/discovery/keys`
   * (jwks_uri from gitlab openid-configuration)
* Set `OIDC Issuer` to `https://gitlab.example.com`
   * (issuer from gitlab openid-configuration)
* Leave `OIDC Audience` as ''
* Install apache module `libapache2-mod-auth-openidc`
* Configure apache module by creating `/etc/apache2/conf.d/05-foreman-ssl.d/01-foreman-oidc.conf` with following contents.
   * (This assumes foreman is configured by `/etc/apache2/sites-available/05-foreman-ssl.conf` with an `IncludeOptional "/etc/apache2/conf.d/05-foreman-ssl.d/*.conf"` line within.)

```
OIDCProviderMetadataURL https://gitlab.example.com/.well-known/openid-configuration
OIDCRedirectURI https://foreman.example.com/users/extlogin/redirect_uri
OIDCClientID "<application id from gitlab>"
OIDCClientSecret "<secret from gitlab>"
OIDCCryptoPassphrase "exec:/bin/bash -c \"head /dev/urandom | tr -dc A-Za-z0-9 | head -c 32\""
OIDCScope "openid profile email"
# Provides HTTP_OIDC_ID_TOKEN_PAYLOAD header variable.
OIDCPassIDTokenAs "payload"
# Provides HTTP_OIDC_USERINFO_JSON header variable.
OIDCPassUserInfoAs "json"

<Location /users/extlogin>
  AuthType openid-connect
  Require valid-user
</Location>
```

stejskalleos commented:
[test foreman]

stejskalleos commented on app/services/sso/openid_connect.rb:
```diff
@@ -1,4 +1,12 @@
+  #     OIDCPassIDTokenAs "payload" # Provides HTTP_OIDC_ID_TOKEN_PAYLOAD header variable.
+  #     OIDCPassUserInfoAs "json" # Provides HTTP_OIDC_USERINFO_JSON header variable.
+  # OIDC scope should include profile and email for foreman account creation e.g.
+  #  OIDCScope "openid profile email"
```
Nit: padding

stejskalleos commented on app/services/sso/openid_connect.rb:
```diff
@@ -8,13 +16,15 @@ def available?
+      return false unless (payload = id_token_payload)
+      return false unless (user = find_or_create_user_from_payload(payload))
 
+      logger = Foreman::Logging.logger('app')
```
You can use `Rails.logger.info|warn|error`

stejskalleos commented on app/services/sso/openid_connect.rb:
```diff
@@ -46,47 +56,80 @@ def logout_url
+        userinfo = JSON.parse(http_oidc_userinfo_json)
+        payload = payload.merge(userinfo)
+      end
+      logger = Foreman::Logging.logger('app')
```
You can use `Rails.logger.info|warn|error`

stejskalleos commented on app/services/sso/openid_connect.rb:
```diff
@@ -46,47 +56,80 @@ def logout_url
+      # Detect keycloak attrs, falling back to gitlab attrs.
+      attr_login = payload.has_key?('preferred_username') ? payload['preferred_username'] : payload['nickname']
+      attr_firstname = payload.has_key?('given_name') ? payload['given_name'] : payload['name']
+      attr_lastname = payload.has_key?('family_name') ? payload['family_name'] : ''
```
This could be:
```ruby
attr_login = payload.fetch('preferred_username', payload['nickname'])
attr_firstname = payload.fetch('given_name', payload['name'])
attr_lastname = payload.fetch('family_name', '')
```

stejskalleos reviewed (commented):
Thanks  @j-m-harris for the update and instructions how to update setup the Openid SSO.
It works very well, didn't have any issues, just few minor comments:
* We should update the documentation, can you post the link to documentation PR?
* When I was trying to authenticate admin user, I was always redirected to the `Any Organization` and `Any Location`, despite the fact that I did have default org/loc set to specific org/loc. Not sure if it is a problem just this SSO or it's a bug for all of them
* Rubocop is not happy:
```
Offenses:
test/unit/sso/openid_connect_test.rb:58:9: C: Layout/FirstHashElementIndentation: Use 2 spaces for indentation in a hash, relative to the first position after the preceding left parenthesis.
        'HTTP_OIDC_ACCESS_TOKEN' => '593fb2dfc385a1550ec43ddd60edbf6c',
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
test/unit/sso/openid_connect_test.rb:60:7: C: Layout/FirstHashElementIndentation: Indent the right brace the same as the first position after the preceding left parenthesis.
      })
      ^
test/unit/sso/openid_connect_test.rb:[13](https://github.com/theforeman/foreman/runs/7316919959?check_suite_focus=true#step:4:14)6:9: C: Layout/FirstHashElementIndentation: Use 2 spaces for indentation in a hash, relative to the first position after the preceding left parenthesis.
        'HTTP_OIDC_ACCESS_TOKEN' => '593fb2dfc385a1550ec43ddd60edbf6c',
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
test/unit/sso/openid_connect_test.rb:139:7: C: Layout/FirstHashElementIndentation: Indent the right brace the same as the first position after the preceding left parenthesis.
      })
```

stejskalleos commented:
@j-m-harris any updates?

stejskalleos commented:
I'm closing the PR, it's been without any update since the Oct 3, 2022.
